### PR TITLE
Skip 2PC for readonly connections in a transaction

### LINKED ISSUE #3250: 2PC should skip read-only participants?
## Scenario
Multiple workers are involved in a global transaction, some of them are read-only. When the transaction commits, all participants go through prepare and commit phases. 

However, the read-only participant has nothing to prepare, since it does not change the state of database.

## Question
Cloud Citus  optimizes 2PC to skip read-only participants, which may create great contributions in some circumstance?

### PR DESCRIPTION
**Pull request by naisila**
DESCRIPTION: Skip 2PC for readonly connections in a transaction

Check whether the connection performed a write by looking at `hadDDL` and `hadDML` variables of `Connection Reference` struct through the available `ConnectionModifiedPlacement()` function.

TODO:

- [x] fix test output by determining order of NOTICEs and DETAILs

Fixes #3250 , #4510

### CODE REVIEW

File: src/backend/distributed/transaction/remote_transaction.c [hunk 1]
```diff
@@ -782,8 +783,24 @@ CoordinatedRemoteTransactionsPrepare(void)
...
 
-		StartRemoteTransactionPrepare(connection);
-		connectionList = lappend(connectionList, connection);
+		/*
+		 * Check if any DML or DDL is executed over the connection on any
+		 * placement/table. If yes, we start preparing the transaction, otherwise
+		 * we skip prepare since the connection didn't perform any write (read-only)
+		 */
+		dlist_iter placementIter;
+		dlist_foreach(placementIter, &connection->referencedPlacements)
```
Comment (SaitTalhaNisanci) [inline src/backend/distributed/transaction/remote_transaction.c:792]: I would suggest adding a field to Connection struct, `hasProcessedWrite` and setting it to true in these places:

https://github.com/citusdata/citus/blob/94bc2703bce352fc5837ca1bda5b8fe65569097d/src/backend/distributed/connection/placement_connection.c#L459

https://github.com/citusdata/citus/blob/94bc2703bce352fc5837ca1bda5b8fe65569097d/src/backend/distributed/connection/placement_connection.c#L464

So basically `connection->hasProcessedWrite = true;` could be there, not `placementConnection->hasProcessedWrite`.

Then here the instead of `dlist_foreach` the check could become:

```c
if (connection->hasProcessedWrite) {
				StartRemoteTransactionPrepare(connection);
				connectionList = lappend(connectionList, connection);
}
```

Would this work?

File: src/backend/distributed/transaction/remote_transaction.c [hunk 2]
```diff
@@ -801,7 +818,23 @@ CoordinatedRemoteTransactionsPrepare(void)
...
 		}
 
-		FinishRemoteTransactionPrepare(connection);
+		/*
+		 * Check if any DML or DDL is executed over the connection on any
+		 * placement/table. If yes, we finish preparing the transaction, otherwise
+		 * we skip prepare since the connection didn't perform any write (read-only)
+		 */
+		dlist_iter placementIter;
+		dlist_foreach(placementIter, &connection->referencedPlacements)
```
Comment (SaitTalhaNisanci) [inline src/backend/distributed/transaction/remote_transaction.c:827]: And here instead of `dlist_foreach` the check could become:

```
if (connection->hasProcessedWrite) {
     FinishRemoteTransactionPrepare(connection);
}
```

File: src/test/regress/sql/multi_transaction_2pc.sql [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+SET citus.shard_replication_factor to 1;
+
+-- check that read-only participants skip prepare
+-- only two of the connections will perform a write (INSERT)
+CREATE TABLE test (a int);
+SELECT create_distributed_table('test', 'a');
+INSERT INTO test SELECT i FROM generate_series(0, 5)i;
+SET citus.log_remote_commands TO ON;
+BEGIN;
+INSERT INTO test VALUES (1);
```
Comment (SaitTalhaNisanci) [inline src/test/regress/sql/multi_transaction_2pc.sql:18]: Instead of having settnig logging the remote commands on and off for this block, you could possibly use `set local citus.log_remote_commands to on;` instead the transaction block, the effects of which will end afer commit/rollback.

Comment (SaitTalhaNisanci) [inline src/test/regress/sql/multi_transaction_2pc.sql:18]: Could possibly add another begin..end block where we insert multiple values in one statement to `test` table and then have a `count(*)`

File: src/backend/distributed/connection/placement_connection.c [hunk 4]
```diff
@@ -32,45 +32,6 @@
 #include "utils/memutils.h"
 
 
-/*
```
Comment (onderkalaci) [inline src/backend/distributed/connection/placement_connection.c:35]: This structure keeps an internal state, and better suited to stay in this file. Can we avoid making this struct public (e.g., not move to the header)? Maybe by accessing the relevant fields via a function?

File: src/backend/distributed/transaction/remote_transaction.c (hunk 1 above)
Comment (onderkalaci) [reply src/backend/distributed/transaction/remote_transaction.c:792]: I'd avoid adding a new flag. Can we use `ConnectionModifiedPlacement()` function?

File: src/backend/distributed/connection/placement_connection.c (hunk 4 above)
Comment (SaitTalhaNisanci) [reply src/backend/distributed/connection/placement_connection.c:35]: If we do the suggestions mentioned here: https://github.com/citusdata/citus/pull/4587/files#r565098083, we won't need to use anything from this struct I think, hence it can go back without introducing any function

Comment (onderkalaci) [reply src/backend/distributed/connection/placement_connection.c:35]: The function already exists: https://github.com/citusdata/citus/blob/master/src/backend/distributed/connection/placement_connection.c#L828-L861

Comment (SaitTalhaNisanci) [reply src/backend/distributed/connection/placement_connection.c:35]: Cool, we can directly use that function instead of adding a new field to connection struct then

**Comment by naisila** [comment]
I set `citus.log_remote_commands` to `off` for the `SELECT` queries in the tests since I didn't manage to display the `NOTICE`s and `DETAIL`s in a particular order. But we know there are more than two connections opened for the 'SELECT' queries (whereas for the write queries there are two connections). Therefore we verify the tests by checking that there are only two `PREPARE TRANSACTION`s in the commit.

File: src/backend/distributed/transaction/remote_transaction.c [hunk 5]
```diff
@@ -801,7 +810,15 @@ CoordinatedRemoteTransactionsPrepare(void)
 			continue;
 		}
 
-		FinishRemoteTransactionPrepare(connection);
+		/*
```
Comment (onderkalaci) [inline src/backend/distributed/transaction/remote_transaction.c:813]: Isn't the above `if (transaction->transactionState != REMOTE_TRANS_PREPARING)` already skips the read-only connections? If so, can we avoid the second call to `ConnectionModifiedPlacement ` as that might be an expensive call.

So, maybe convert it to an assert such as below?

```C
if (transaction->transactionState != REMOTE_TRANS_PREPARING)
{
    Assert (!ConnectionModifiedPlacement(connection))
    continue;
}
```

Also, in general `ConnectionModifiedPlacement` is pretty expensive check, iterating on all references. And this code-path is performance critical, so we'd want to avoid any additional overhead.

Is there a way to avoid the call with maybe (didn't check this idea in depth, just suggesting for further investigation and not to forget :) ):
```C
if (XactModificationLevel >= XACT_MODIFICATION_DATA && ConnectionModifiedPlacement(connection))
{
    ... StartRemoteTransactionPrepare...
}
```

Comment (SaitTalhaNisanci) [reply src/backend/distributed/transaction/remote_transaction.c:813]: Depending on a previous statement seems a bit fragile to me, I understand the performance implication and the benefit of `if (XactModificationLevel >= XACT_MODIFICATION_DATA && ConnectionModifiedPlacement(connection))` but this seems to be complicating the code. Why don't we just have a flag in the connection struct that way we can access that field with 0 cost and use it whenever we need it without having to worry about any performance problems. That seems much simpler to me

Comment (onderkalaci) [reply src/backend/distributed/transaction/remote_transaction.c:813]: `XACT_MODIFICATION_DATA ` is not necessary and do not work anyway. So, ignore that please.


> Why don't we just have a flag in the connection struct that way we can access that field with 0 cost and use it whenever we need it without having to worry about any performance problems. 

I'm not completely against that, that can be an option. I see two aspects of it:

- Having an additional flag which keeps the same state as the others should not be required unless we have a really good reason. Because, it is always prone to bugs in the future (as someone might forget to update one of the flags).
- Do we really need this performance optimization? The most typical source of bugs is that we over engineer/optimize things that are not strictly necessary. **In any case,** we need some performance benchmarks for this PR. For example, do this PR add any performance penalty on HammerDB benchmarks? Overall, I'd prefer to give data-driven decisions here. And the benchmarks I'd do is the ones that require 2PC. Does HammerDB require? I think so. 


So, lets make sure that we run at least HammerDB and pgbench's tpcb-like benchmarks with `ConnectionModifiedPlacement ()` calls in the code vs master.  And, similarly some more tests where we  iterate inside the `ConnectionModifiedPlacement ()`, such as the following on a single thread pgbench:
```
-- set the pool size to 1 and make sure that all
-- placements are accessed over the same conection
SET citus.adaptive_executor_pool_size TO 1;
DELETE FROM table;
```

With this data, we can have a better idea on whether we need this optimizations or not

Comment (SaitTalhaNisanci) [reply src/backend/distributed/transaction/remote_transaction.c:813]: tpcc results:
https://github.com/citusdata/release-test-results/tree/citus_github_push/citusbot_tpcc_benchmark_rg/02_03_2021_1612347049/02_03_2021_1612361561/hammerdb/02_03_2021_1612361561

399K current_branch vs 392K master
seems like no regression.

**Comment by naisila** [comment]
@onderkalaci @SaitTalhaNisanci 

* As Talha said, in HammerDB it seems like there is no regression
399K issue2pc (current branch) vs 392K master
* The general pgbench job has shows no regression as seen in https://github.com/citusdata/release-test-results/blob/citusbot_pgbench_test_resource_group/02_09_2021_1612912666/periodic_job_results/02_09_2021_1612912666/pgbench_results_2021-02-09-21-02_pgbench_default.ini.csv
     * As seen in the above csv, there seemed to be around 1% regression in `select-only` job. I ran select-only job for 1 hour instead of 10 minutes to have a more confident result. As a result, 38442 tps in master vs 38253 tps in issue2pc (less than 0.5% regression). 
* I ran the following on a single thread pgbench in my local machine:
1213 tps in master vs 1313 tps in issue2pc, so seems like no regression here either.
```SQL
-- set the pool size to 1 and make sure that all
-- placements are accessed over the same conection
SET citus.adaptive_executor_pool_size TO 1;
DELETE FROM table;
```
It seems like there is no significant regression to opt for a performance optimization.

File: src/test/regress/sql/multi_transaction_2pc.sql [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+-- check that reads from a reference table don't trigger 2PC
+-- despite repmodel being 2PC
+CREATE TABLE test_reference (b int);
+SELECT create_reference_table('test_reference');
+INSERT INTO test_reference VALUES(1);
+INSERT INTO test_reference VALUES(2);
+BEGIN;
+SET LOCAL citus.log_remote_commands TO ON;
+SELECT * FROM test_reference;
```
Comment (SaitTalhaNisanci) [inline src/test/regress/sql/multi_transaction_2pc.sql:47]: Can you add an `ORDER BY` so that this line doesn't get flaky. If you don't specify the order then there is no guarantee in the order.

**Review by SaitTalhaNisanci — APPROVED** [review]

File: src/test/regress/bin/normalize.sed [hunk 7]
```diff
@@ -200,6 +200,10 @@ s/relation with OID [0-9]+ does not exist/relation with OID XXXX does not exist/
 # normalize storage id of columnar tables
 s/^storage id: [0-9]+$/storage id: xxxxx/g
 
+# normalize PREPARE TRANSACTION and COMMIT PREPARED
```
Comment (onderkalaci) [inline src/test/regress/bin/normalize.sed:203]: PREPARE TRANSACTION / COMMIT PREPARED are not intended to be printed. If we add this rule, we'll allow any further tests to print these commands, which doesn't sound like a good idea to me.

 I think we should add the regression tests similar to this file: https://github.com/citusdata/citus/blob/21fd2e2c92bc462aded1ad5f39dbdeb25ea198f0/src/test/regress/sql/multi_transaction_recovery.sql

Where we disable 2PC recovery, then execute the commands and check `pg_dist_transaction`.


Note that, if we do that, we cannot run the the test file in parallel with any other test as they might become flaky because of disabling 2PC recovery. 

We could even move these tests there.

**Review by onderkalaci — APPROVED** [review]
> -- set the pool size to 1 and make sure that all
-- placements are accessed over the same conection
SET citus.adaptive_executor_pool_size TO 1;
DELETE FROM table;

This should have been:
```SQL
BEGIN;
INSERT INTO one shard;
INSERT INTO other shard over another connection;
SELECT count(*) FROM table
```

Though, I've done this and no overhead is observed. After considering my testing suggestion, lets merge this as @marcocitus also suggested on #4792, and it simplifies reasoning about other cases.

File: src/test/regress/expected/multi_transaction_recovery.out [hunk 8]
```diff
@@ -331,6 +331,104 @@ SELECT count(*) FROM pg_dist_transaction;
...
+
+INSERT INTO test_2pcskip SELECT i FROM generate_series(0, 5)i;
+SELECT recover_prepared_transactions();
+ recover_prepared_transactions
+---------------------------------------------------------------------
+                             0
+(1 row)
+
+BEGIN;
+-- these inserts use two connections
```
Comment (onderkalaci) [inline src/test/regress/expected/multi_transaction_recovery.out:352]: This information is vague. Can we do something like to ensure that these values goes to different shards on different workers?:

```
SELECT count(DISTINCT nodeport) FROM pg_dist_shard_placement WHERE shardid IN (get_shard_id_for_distribution_column('test', 6),get_shard_id_for_distribution_column('test', 7));
``` 

Otherwise if shard count/worker count changes, this test may not work as intended

File: src/test/regress/expected/multi_transaction_recovery.out [hunk 9]
```diff
@@ -331,6 +331,104 @@ SELECT count(*) FROM pg_dist_transaction;
...
+ recover_prepared_transactions
+---------------------------------------------------------------------
+                             0
+(1 row)
+
+-- check that read-only participants skip prepare
+-- only two of the connections will perform a write (INSERT)
+BEGIN;
+-- this insert uses two connections
+INSERT INTO test_2pcskip SELECT i FROM generate_series(8, 10)i;
```
Comment (onderkalaci) [inline src/test/regress/expected/multi_transaction_recovery.out:379]: can we use the same values as we used before, 6 and 7 to make sure that these two values are on different shards/workers?

File: src/test/regress/expected/multi_transaction_recovery.out (hunk 8 above)
Comment (naisila) [reply src/test/regress/expected/multi_transaction_recovery.out:352]: Good idea

Comment (onderkalaci) [reply src/test/regress/expected/multi_transaction_recovery.out:352]: > SELECT count(DISTINCT nodeport) FROM pg_dist_shard_placement WHERE shardid IN (get_shard_id_for_distribution_column('test', 6),get_shard_id_for_distribution_column('test', 7));

test should be `test_2pcskip`, that is the table name if you are not familiar with this UDF.
